Tests/DatadogTests/DatadogPrivate/URLSessionSwizzlerTests.swift

```diff
@@ -0,0 +1,410 @@
+        tearDownTracedSession()
+    }
+
+    func testSwizzlingTheSwizzled() {
+        let interceptor: RequestInterceptor = { $1 }
+        let observer: TaskObserver = { _, _ in }
+        let session = URLSession(configuration: .default)
+        defer {
+            try! URLSessionSwizzler.unswizzle(session, disposeDynamicClass: true)
+        }
+        do {
+            for _ in 0...5 {
+                try URLSessionSwizzler.swizzle(
+                    session,
+                    requestInterceptor: interceptor,
+                    taskObserver: observer
+                )
+            }
+        } catch {
+            XCTAssertNil(error)
+        }
+    }
+
+    func testSwizzleMultipleTimes() {
+        /*
+         1. call Datadog.trace(session) multiple times consecutively
+         2. expect the first call to take longer than the last ones
+         NOTE: this tests dynamic class cache in URLSessionSwizzler.m
+         */
+        let session1 = URLSession(configuration: .default)
+        let session2 = URLSession(configuration: .default)
+        let session3 = URLSession(configuration: .default)
+
+        do {
+            try URLSessionSwizzler.swizzle(
+                session1,
+                requestInterceptor: { $1 },
+                taskObserver: { _, _ in },
+                enforceNewClassConfiguration: true
+            )
+            try URLSessionSwizzler.swizzle(
+                session2,
+                requestInterceptor: { $1 },
+                taskObserver: { _, _ in },
+                enforceNewClassConfiguration: false
+            )
+            try URLSessionSwizzler.swizzle(
+                session3,
+                requestInterceptor: { $1 },
+                taskObserver: { _, _ in },
+                enforceNewClassConfiguration: true
+            )
+            XCTFail("session3 swizzling with enforceDynamicClassCreation should throw!")
+        } catch {
+            let className1 = NSStringFromClass(object_getClass(session1)!)
+            let className2 = NSStringFromClass(object_getClass(session2)!)
+            let className3 = NSStringFromClass(object_getClass(session3)!)
+
+            XCTAssertTrue(className1.contains("_Datadog"))
+            XCTAssertEqual(className1, className2)
+            XCTAssertFalse(className3.contains("_Datadog"))
+            XCTAssertNotNil(error)
+        }
+
+        // tearDownTracedSession(session1/2) would result in EXC_BAD_ACCESS
+        // Because tearDownTracedSession removes dynamic class by default
+        // If we tearDown one of those instances, the other would be class-less -> EXC_BAD_ACCESS
+        try! URLSessionSwizzler.unswizzle(
+            session1,
+            disposeDynamicClass: false
+        )
+        try! URLSessionSwizzler.unswizzle(
+            session2,
+            disposeDynamicClass: true
+        )
+        tearDownTracedSession(session3)
+    }
+
+//    func testConcurrentSwizzling() {
```

Comment:
Let's uncomment this? Ideally, the thread safety shouldn't be considered on `URLSessionSwizzler`, but on `Swizzler`, so we do not have to repeat the thread-safety logic in all swizzlers. Is it doable?

Reply:
I had 2 issues when I tried to add thread-safety to `Swizzler` instead of `URLSessionSwizzler`

First problem was, `dispatch_a/sync` and `(NSError **)error` don't go well together because of ARC/memory management. Compiler complains about it, although there is a certain syntax to make it work it didn't feel safe to me

But tha main issue was that dispatching one block and getting things done in one go vs. dispatching multiple blocks
As `Swizzler` is a set of methods and `XXXSwizzler`s are usually one or two methods, I prefer syncing the whole operation in one go at `XXXSwizzler` level.
It came simpler, safer and faster. Downside is having to write one more test case for each `XXXSwizzler`, but IMO this is fine